runtime/lua/vscode-neovim/viewport.lua

```diff
@@ -0,0 +1,62 @@
+local M = {}
+
+local api, fn = vim.api, vim.fn
+
+M.event_group = api.nvim_create_augroup("vscode-neovim.viewport", { clear = true })
+M.viewport_changed_ns = api.nvim_create_namespace("vscode-neovim.viewport.changed")
+
+---@class WinView All positions are 0-based
+---@field lnum integer
+---@field col integer
+---@field coladd integer
+---@field curswant integer
+---@field topline integer
+---@field topfill integer
+---@field leftcol integer
+---@field skipcol integer
+---@field winid integer
+
+---@return WinView
+local function get_winview()
+  local view = fn.winsaveview()
+  view.lnum = view.lnum - 1
+  view.topline = view.topline - 1
+  view.winid = api.nvim_get_current_win()
+  return view
+end
+
+local function setup_viewport_changed()
+  ---@type table<integer, WinView>
+  local view_cache = {}
+
+  api.nvim_set_decoration_provider(M.viewport_changed_ns, {
+    on_win = function()
```

Comment:
Just for my education, what's the difference between this and the `WinScrolled` autocmd?

Reply:
WinScrolled won't be triggered under cmdline. We also need timely and accurate viewport information https://github.com/neovim/neovim/issues/28800#issuecomment-2117611542.

Reply:
Huh, ok.

This comment befuddles me a little bit, but the reason for your change makes sense :)

> Well yes the cursor position in the current window changes for incsearch but the cursor position on the grid does not and should not(implicitly). What problem does this cause?

Reply:
- "grid cursor" refers to the **visible** cursor position.
- One grid includes not only the window viewport but also components like cmdline, statuscolumn, winbar, and so on. However, vscode-neovim doesn't (can't) need these components, we only dealing with the window viewport.

Reply:
You may notice that in the cursor manager, we only use `grid_cursor_goto` to detect changes in cursor position, but we actually use data from the viewport. With `ext_cmdline` enabledthe visual cursor position is managed by the UI, so there is no `grid_cursor_goto` event (In Nvim0.10).